File: 73. Set Matrix Zeroes.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        O(m+n) space and O(mn) time
        """
        row = set()
        column = set()
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j] == 0:
                    row.add(i)
                    column.add(j)
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if i in row or j in column:
                    matrix[i][j] = 0

    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        O(1) space O(mn) time, is_col is for matrix[0][0] to see whether it is both column and row or just row
        """
        is_col = False
        R = len(matrix)
        C = len(matrix[0])
        for i in range(R):
            if matrix[i][0] == 0:
                is_col = True
            for j in range(1, C):
                if matrix[i][j] == 0:
                    matrix[0][j] = 0
                    matrix[i][0] = 0
        for i in range(1,R):
            for j in range(1,C):
                if matrix[0][j] == 0 or matrix[i][0] == 0:
                    matrix[i][j] = 0
        if matrix[0][0] == 0:
            for j in range(C):
                matrix[0][j] = 0  
        if is_col:
            for i in range(R):
                matrix[i][0] = 0

File: 73. Set Matrix Zeroes.java
class Solution {
    public void setZeroes(int[][] matrix) {
        if (matrix==null || matrix.length==0) return;
        boolean fromRow=false, fromCol=false;
        int m=matrix.length, n=matrix[0].length;
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (matrix[i][j] == 0) {
                    if (i==0) fromRow = true;
                    if (j==0) fromCol = true;
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        for (int i=1; i<m; i++) {
            if (matrix[i][0] == 0) {
                for (int j=1; j<n; j++) {
                    matrix[i][j] = 0;
                }
            }
        }
        for (int j=1; j<n; j++) {
            if (matrix[0][j] == 0) {
                for (int i=1; i<m; i++) {
                    matrix[i][j] = 0;
                }
            }
        }
        if (fromRow) {
            for (int j=1; j<n; j++) {
                matrix[0][j] = 0;
            }
        }
        if (fromCol) {
            for (int i=1; i<m; i++) {
                matrix[i][0] = 0;
            }
        }
    }
}